src/cmd/edit.rs

```diff
@@ -0,0 +1,267 @@
+use crate::cmd::{Edit, Run};
+use crate::db::{Database, DatabaseFile, Epoch, Rank};
+use crate::util::resolve_path;
+use crate::{config, util};
+use anyhow::Result;
+use std::fmt::Write as FmtWrite;
+use std::path::PathBuf;
+
+use dialoguer::Input;
+use edit::edit;
+
+const HEADER: &str = "\
+# Blank lines and lines prepended with '#' are ignored; Line order is insignificant
+# last_accessed,rank,path
+";
+
+enum ValidationResult {
+    Success,
+    Retry,
+    Exit,
+}
+
+struct Problem {
+    line_number: usize,
+    text: String,
+}
+
+struct Problems {
+    warnings: Vec<Problem>,
+    errors: Vec<Problem>,
+}
+
+impl Problem {
+    fn new(line_number: usize, text: String) -> Self {
+        Problem { line_number, text }
+    }
+}
+
+impl Run for Edit {
+    fn run(&self) -> Result<()> {
+        while let Some(db_edits) = get_db_edits()? {
+            let data_dir = config::data_dir()?;
+            let mut db_file = DatabaseFile::new(data_dir);
+            let mut db = db_file.open()?;
+            db.clear();
+            let problems = get_problems(&mut db, db_edits);
+            let result = handle_problems(problems);
+            match result {
+                ValidationResult::Success => {
+                    db.save()?;
+                    return Ok(());
+                }
+                ValidationResult::Exit => break,
+                ValidationResult::Retry => continue,
+            }
+        }
+        println!("Zoxide database not altered");
+        Ok(())
+    }
+}
+
+fn get_db_edits() -> Result<Option<String>> {
+    let data_dir = config::data_dir()?;
+    let mut db = DatabaseFile::new(data_dir);
+    let mut db = db.open()?;
+    let mut stream = db.stream(util::current_time().unwrap());
```

Comment:
Streams run a bunch of filtering on the DB which we don't want to do here (we want to show everything in the DB, valid or not). I'd suggest just accessing the `dirs` field directly.